src/hotspot/share/gc/parallel/mutableSpace.cpp

```diff
@@ -235,7 +235,13 @@ void MutableSpace::oop_iterate(OopIterateClosure* cl) {
 void MutableSpace::object_iterate(ObjectClosure* cl) {
   HeapWord* p = bottom();
   while (p < top()) {
-    cl->do_object(cast_to_oop(p));
+    oop obj = cast_to_oop(p);
+    // When promotion-failure occurs during Young GC, eden/from space is not cleared,
+    // so we can encounter objects with "forwarded" markword.
+    // They are essentially dead, so skipping them
+    if (!obj->is_forwarded() || obj->forwardee() == obj) {
```

Comment:
Self-forwarded ones should already be handled by `PSPromotionManager::restore_preserved_marks`, so the second condition will always be false.